src/EntityFramework.SqlServer/SqlGen/SqlFunctionCallHandler.cs

```diff
@@ -1285,49 +1285,63 @@ private static void AppendConvertToVarchar(SqlGenerator sqlgen, SqlBuilder resul
 
         // <summary>
         // TruncateTime(DateTime X)
-        // PreKatmai:    TRUNCATETIME(X) => CONVERT(DATETIME, CONVERT(VARCHAR(255), expression, 102),  102)
-        // Katmai:    TRUNCATETIME(X) => CONVERT(DATETIME2, CONVERT(VARCHAR(255), expression, 102),  102)
+        // PreKatmai:    TRUNCATETIME(X) => DATEADD(d, DATEDIFF(d, 0, expression), 0)
+        // Katmai:       TRUNCATETIME(X) => CAST(CAST(expression AS DATE) as DATETIME2)
```

Comment:
There appears to be no significant perf advantage advantage in the Katmai translation over the PreKatmai translaiton. I tested it with:

``` sql
declare @expression datetime;
declare @ret datetime2;

set @expression = CAST('1969/08/22 10:10' as DATETIME2);


DECLARE @Iteration INT
SET @Iteration = 1
WHILE @Iteration <= 100000000
BEGIN

	--SET @ret = CONVERT(DATETIME, convert(varchar(255), @expression, 102), 102); -- 92 seconds
	--SET @ret = DATEADD(d, DATEDIFF(d, 0, @expression), 0); -- 55 seconds
	SET @ret = CAST(CAST(@expression AS DATE) as DATETIME2); -- 54 seconds

    SET @Iteration = @Iteration + 1

END

print @ret
print @iteration
```
I would suggest we simplify.

Reply:
Tested this myself and got very similar results with your test, but found much less CPU time was used on the Katmai translation based on the following script:

```
-- Setup 
CREATE TABLE #test (Value datetime2)

SET NOCOUNT ON;

DECLARE @Iteration INT
SET @Iteration = 1
WHILE @Iteration <= 10000000
BEGIN
    INSERT INTO #Test
    SELECT DATEADD(mi, @Iteration, CAST('1969/08/22 10:10' as DATETIME2))


    SET @Iteration = @Iteration + 1
    IF CAST((@Iteration / 100000) as int) * 100000 = @Iteration
       print @iteration

END


-- Test
SET STATISTICS TIME ON;
SET STATISTICS IO ON;

-- Table '#Test__...'. Scan count 9, logical reads 22076, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
--
-- SQL Server Execution Times:
--   CPU time = 1423 ms,  elapsed time = 221 ms.
SELECT 1 from #test where CAST(CAST(value AS DATE) as DATETIME2) = CAST('2019-01-05' as datetime2)

--Table '#Test__...'. Scan count 9, logical reads 22076, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
--
-- SQL Server Execution Times:
--   CPU time = 2844 ms,  elapsed time = 379 ms.
SELECT 1 from #test where DATEADD(d, DATEDIFF(d, 0, Value), 0) = CAST('2019-01-05' as datetime2)


SET STATISTICS TIME OFF;
SET STATISTICS IO OFF;


-- Cleanup
DROP TABLE #test;
```

Based on that test it appears that it might be as much as 50% less cpu-intensive to use the Katmai version over the Pre-Katmai version.  I don't particular have an explanation for why the results are different from your test's results elapsed-time wise.

Reply:
Very cool! I can confirm your findings.